whd-zephyr-integration/cyw43xxx_drv.c

```diff
@@ -0,0 +1,720 @@
+#include <zephyr/net/net_if.h>
+#include <zephyr/device.h>
+#include <soc.h>
+
+#include <zephyr/net/wifi_mgmt.h>
+#include <zephyr/net/net_offload.h>
+#include <zephyr/net/net_event.h>
+#include <zephyr/net/net_mgmt.h>
+
+#include <cybsp_wifi.h>
+#include <cyhal.h>
+#include <whd_types.h>
+#include <whd_events_int.h>
+#include <whd_buffer_api.h>
+#include <cy_network_buffer.h>
+#include <cybsp_wifi.h>
+#include <cybsp.h>
+
+#ifndef CYW43XXX_TX_PACKET_POOL_COUNT
+#define CYW43XXX_TX_PACKET_POOL_COUNT (10)
+#endif
+
+#ifndef CYW43XXX_RX_PACKET_POOL_COUNT
+#define CYW43XXX_RX_PACKET_POOL_COUNT (10)
+#endif
+
+#ifndef CYW43XXX_PACKET_POOL_SIZE
+#define CYW43XXX_PACKET_POOL_SIZE (1600)
+#endif
+
+#define CYW43XXX_PACKET_POOL_COUNT (CYW43XXX_TX_PACKET_POOL_COUNT + CYW43XXX_RX_PACKET_POOL_COUNT)
+
+/* This macro is copy of NET_BUF_POOL_FIXED_DEFINE with aligning net_buf_data_##_name
+ * WHD requires that network buffers is aligned, NET_BUF_POOL_FIXED_DEFINE does not
+ * guarantees aligned.
+ */
+#define NET_BUF_POOL_FIXED_DEFINE_ALIGN(_name, _count, _data_size, _ud_size, _destroy)             \
+	_NET_BUF_ARRAY_DEFINE(_name, _count, _ud_size);                                            \
+	static uint8_t __noinit net_buf_data_##_name[_count][_data_size] __net_buf_align;          \
+	static const struct net_buf_pool_fixed net_buf_fixed_##_name = {                           \
+		.data_size = _data_size,                                                           \
+		.data_pool = (uint8_t *)net_buf_data_##_name,                                      \
+	};                                                                                         \
+	static const struct net_buf_data_alloc net_buf_fixed_alloc_##_name = {                     \
+		.cb = &net_buf_fixed_cb,                                                           \
+		.alloc_data = (void *)&net_buf_fixed_##_name,                                      \
+	};                                                                                         \
+	static STRUCT_SECTION_ITERABLE(net_buf_pool, _name) = NET_BUF_POOL_INITIALIZER(            \
+		_name, &net_buf_fixed_alloc_##_name, _net_buf_##_name, _count, _ud_size, _destroy)
+
+#define CYW43XXX_WAIT_SEMA_MS (30 * 1000)
+
+/* Allocate network pool */
+NET_BUF_POOL_FIXED_DEFINE_ALIGN(cyw43xxx_pool, CYW43XXX_PACKET_POOL_COUNT,
+				CYW43XXX_PACKET_POOL_SIZE, 0, NULL);
+
+static uint16_t ap_event_handler_index = 0xFF;
+
+/* Use global iface pointer to support any Ethernet driver */
+/* necessary for wifi callback functions */
+static struct net_if *cyw43xxx_wifi_iface;
+
+struct cyw43xxx_wifi_runtime {
+	struct net_if *iface;
+	bool second_interface_init;
+	bool is_ap_up;
+	bool is_sta_connected;
+	uint8_t mac_addr[6];
+	scan_result_cb_t scan_rslt_cb;
+	whd_ssid_t ssid;
+	whd_scan_result_t scan_result;
+	struct k_sem sema_common;
+	struct k_sem sema_scan;
+	uint8_t frame_buf[NET_ETH_MAX_FRAME_SIZE];
+#if defined(CONFIG_NET_STATISTICS_WIFI)
```

Comment:
what is the significance of framebuf size when you are using Wifi-EHD integrated with SDIO?

Reply:
@npal-cy can you share your thoughts?

Reply:
let me check this one this my team, will update.

Reply:
Sure, Can you also verify the value of WHD_BUS_SDIO_MAX_BACKPLANE_TRANSFER_SIZE which is set to 1536 currently. we think it should be 1568?

Reply:
> Can you also verify the value of WHD_BUS_SDIO_MAX_BACKPLANE_TRANSFER_SIZE which is set to 1536 currently. we think it should be 1568?

Hi @prayassamriya, could you please provide more details, how to reproduce the issue with WHD_BUS_SDIO_MAX_BACKPLANE_TRANSFER_SIZE ? What sample are you using?